Fix URL Redirection in OAuth2/OpenID/SAML

br41nslug opened the pull request:
## Scope

What's changed:

Adds an allow list for OpenID/Oauth login redirects to prevent redirecting to arbitrary domains. This is not the callback redirect from SSO but the Directus redirect query parameter which redirects after successful login.

## Potential Risks / Drawbacks

- This will be a breaking change for anyone relying on this to redirect to an external domain. But can be resolved by adding that to the redirect allow list.

## Review Notes / Questions

- New config option is `AUTH_<PROVIDER>_REDIRECT_ALLOW_LIST`

br41nslug committed:
copied url whitelist checking

br41nslug committed:
Check redirect urls before redirecting to external sso

br41nslug committed:
Added custom env var

br41nslug committed:
extracted redirect allow list check to util function

br41nslug committed:
Added breaking change note

br41nslug committed:
Added new env var to config reference

paescuj committed:
Format files

paescuj committed:
Merge branch 'main' into advisory-fix-1

paescuj committed:
Url -> URL

br41nslug committed:
Merge branch 'main' into advisory-fix-1

br41nslug committed:
resolved env package merge conflict

br41nslug committed:
Merge branch 'advisory-fix-1' of https://github.com/directus/directus-ghsa-fr3w-2p22-6w7p into advisory-fix-1

br41nslug committed:
added basic tests for is-url-allowed

br41nslug committed:
added basic login redirect tests

br41nslug committed:
updated tests and strictness of domain checking

br41nslug committed:
updated test env

br41nslug committed:
added extra tests

br41nslug committed:
added redirect check to SAML

br41nslug committed:
prettier

br41nslug committed:
fixed typo

rijkvanzanten committed:
Merge branch 'main' into fix-sso-redirection

br41nslug committed:
Merge branch 'main' into fix-sso-redirection

br41nslug committed:
prettier

br41nslug committed:
Merge branch 'main' into fix-sso-redirection

azrikahar commented on api/src/auth/drivers/oauth2.ts:
```diff
@@ -298,15 +300,16 @@ export function createOAuth2AuthRouter(providerName: string): Router {
-					issuer: 'directus',
-				},
-			);
+			if (isLoginRedirectAllowed(redirect, providerName)) {
```
```suggestion
			if (isLoginRedirectAllowed(redirect, providerName) === false) {
```

azrikahar commented on docs/releases/breaking-changes.md:
```diff
@@ -13,6 +13,17 @@ these to a minimum, but rest assured we only make them with good reason.
+the `?redirect=http://example.com/login` query parameter then you'll need to add this URL to the
+`AUTH_<PROVIDER>_REDIRECT_ALLOW_LIST`.
+
+# `AUTH_<PROVIDER>_REDIRECT_ALLOW_LIST` accepts a comma separated list of URLs (which include the path when comparing).
```
Not 100% sure, but maybe this wasn't meant to be a h1 tag, perhaps as an info or tip block instead?

paescuj committed:
Specify version in breaking changes docs & remove h1

Co-authored-by: Azri Kahar <42867097+azrikahar@users.noreply.github.com>

paescuj committed:
Fix allowed check

Co-authored-by: Azri Kahar <42867097+azrikahar@users.noreply.github.com>

paescuj committed:
Merge branch 'main' into fix-sso-redirection

paescuj committed:
Clean-up

- Fix PUBLIC_URL comparision if ALLOW_LIST is defined
- Add positive test cases & fix tests
- Docs

paescuj committed:
Add changeset
